[native] Update config properties registration

yhwang opened the pull request:
## Description
Consolidate config properties registration and declaration.

fix: #21345

## Motivation and Context
Config properties declaration and registration are in different places, the declaration is in the `Configs.h` and the registration is in the `Configs.cpp`. It's easy to forget to register the new config property when adding a new one.

## Impact
Refactor the code to make sure every config property is registered.

## Test Plan
Manually test the properties that are not registered and make sure they are not listed as `Unregistered properties` in
the log while launching the native worker

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== NO RELEASE NOTE ==
```

mbasmanova commented:
@yhwang Thank you for the refactoring. Would you add [native] prefix to commit message and PR title?

> Manually test the properties that are not registered and make sure they are not listed as Unregistered properties in
the log while launching the native worker

Would it make sense to introduce a mode where unregistered property results in a hard failure? 

CC: @amitkdutta

yhwang commented:
> Would it make sense to introduce a mode where unregistered property results in a hard failure?

Sounds good to me. Let me check the existing mechanism and see how I can add a mode to do that. Or you can point me to where the relevant code is. If the change is small, then I will include that in this PR. Otherwise, I would have another PR to address that. thought?

mbasmanova commented on presto-native-execution/presto_cpp/main/common/Configs.h:
```diff
@@ -21,10 +21,418 @@
 namespace facebook::presto {
 
+#define SYS_CONF_BASE_PROPS(V)                                                \
+  /* Setting this to 'true' makes configs modifiable via server operations.*/ \
```
I'm not familiar with this code, but I wonder where did these come from? In general, comments on public methods / variables should use ///

yhwang replied:
Yea, I notice the `///` for the general use. However, inside a macro, it's actually a very long string. Therefore, not able to use `///` for the comment. It will comment out all of the following lines.

mbasmanova commented:
> Let me check the existing mechanism and see how I can add a mode to do that. Or you can point me to where the relevant code is. I

I'm not familiar with this code. @amitkdutta @spershin Amit, Sergey, do you have a pointer?

>  If the change is small, then I will include that in this PR. Otherwise, I would have another PR to address that. thought?

A separate PR is probably better.

amitkdutta commented:
> > Would it make sense to introduce a mode where unregistered property results in a hard failure?
> 
> Sounds good to me. Let me check the existing mechanism and see how I can add a mode to do that. Or you can point me to where the relevant code is. If the change is small, then I will include that in this PR. Otherwise, I would have another PR to address that. thought?

@yhwang Here is the code where config properties are checked if the binary supports it https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/common/Configs.cpp#L54
We can add a flag or config for hard failure, making it false by default.

yhwang commented:
rebased the code to solve the conflict which is the newly added config: `task.max-partial-aggregation-memory`

yhwang commented:
fixed the conflict and rebased again.

@aditi-pandit would you help to review this PR? I guess if anyone adds/updates/deletes the config properties, I have to rebase this PR. Not tried to push the merge but hope not to resolve more conflicts.

aditi-pandit commented on presto-native-execution/presto_cpp/main/common/Configs.h:
```diff
@@ -21,10 +21,431 @@
+  /* Setting this to 'true' makes configs modifiable via server operations.*/ \
+  V(BOOL, kMutableConfig, "mutable-config", false)
+
+#define SYS_CONF_PROPS(V)                                                         \
```
Can you retain the original order of properties so that its easier to check that are all still there ?

yhwang replied:
I thought about that but I use alphanumeric and data type as the order at the end. My thought is that it's easier to sort through the properties and I assume people won't care about the grouping and ordering that much at the end of the day.

Let me change back to the original order if you think the original order is meaningful.

BTW, I compared the before and after changes manually, and I am sure I have the whole properties there :-)

yhwang replied:
updated the code to use the original order and rebased the code to solve the conflict.

majetideepak commented on presto-native-execution/presto_cpp/main/common/Configs.h:
```diff
@@ -21,10 +21,433 @@
 
 namespace facebook::presto {
 
+#define SYS_CONF_BASE_PROPS(V)                                                \
```
@yhwang What is the motivation to define this preprocessor? Why refactor this code? Thanks!

aditi-pandit commented on presto-native-execution/presto_cpp/main/common/Configs.cpp:
```diff
@@ -12,8 +12,8 @@
 
-#include "presto_cpp/main/common/ConfigReader.h"
 #include "presto_cpp/main/common/Configs.h"
+#include "presto_cpp/main/common/ConfigReader.h"
```
Nit : The headers have to be in alphabetical order.

yhwang replied:
Not sure why I changed this, revered it. thanks.

aditi-pandit reviewed (commented):
Minor comment

aditi-pandit commented:
@yhwang : Please can you rebase this PR. 

Will work on final reviews on this with priority.
